std-bits/table/src/main/java/org/enso/table/format/xlsx/Reader.java

```diff
@@ -187,32 +198,263 @@ private static Table read_table(
     return new Table(columns);
   }
 
-  private static Object getCellValue(Cell cell, Function<LocalDate, Value> mkDate) {
-    return getCellValue(cell, cell.getCellType(), mkDate);
+  private static Table readSheetToTable(
+      Workbook workbook, int sheetIndex, Range range, int skipRows, int rowCount) {
+    Sheet sheet = workbook.getSheetAt(sheetIndex);
+
+    // Row Range
+    int firstRow = sheet.getFirstRowNum() + 1;
+    int lastRow = sheet.getLastRowNum() + 1;
+    int startRow = (range == null || range.isWholeColumn() ? 1 : range.getTopRow()) + skipRows;
+    int endRow = range == null || range.isWholeColumn() ? lastRow : range.getBottomRow();
+
+    // Columns
+    int startCol = (range == null || range.isWholeRow() ? 1 : range.getLeftColumn());
+    int endCol = (range == null || range.isWholeRow() ? -1 : range.getRightColumn());
+    int size = Math.min(rowCount, endRow - startRow + 1);
+    List<Builder> builders =
+        endCol == -1
+            ? new ArrayList<>()
+            : IntStream.range(startCol, endCol + 1)
+                .mapToObj(i -> new InferredBuilder(size))
+                .collect(Collectors.toList());
+
+    // Read Cell Data
+    int row = startRow;
+    while (row <= endRow && (row - startRow) < rowCount) {
+      if (row < firstRow || row > lastRow) {
+        builders.forEach(b -> b.append(null));
```

Comment:
isn't this going to lead to NPEs later?

Reply:
Nope, in polyglot calls `null` is converted to `Nothing` when crossing the boundary (and IIRC, vice-versa).

NPEs in Enso can happen only if the `null` leaks internally, as in the recursive definition bug.

The Table library is on the same 'level' as any user-provided library, so it benefits from the polyglot translation of nulls.